File: rag/src/storage/vector-store.js
/**
 * vector-store.js — 向量存储抽象层
 *
 * 职责：
 * - 为 chunk 提供向量存储与相似度检索
 * - V1 实现：内存存储 + 余弦相似度（适用于小规模知识库）
 * - 接口设计为可替换，后续可对接 Qdrant/Chroma
 *
 * 设计原则：
 * - 上层代码通过相同接口调用，不依赖具体实现
 * - 后续只需替换此文件（或切换配置）即可升级到 Qdrant
 */

import { existsSync, mkdirSync, writeFileSync, readFileSync } from 'fs';
import { join, resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = resolve(__dirname, '../../..');
const VECTORS_DIR = join(PROJECT_ROOT, 'rag/data/vectors');

let vectors = [];  // [{chunk_id, vector}]
let loaded = false;

/**
 * 加载持久化的向量
 */
function loadVectors() {
  if (loaded) return;
  const filePath = join(VECTORS_DIR, 'vectors.json');
  try {
    if (existsSync(filePath)) {
      const data = readFileSync(filePath, 'utf-8');
      vectors = JSON.parse(data);
      console.log(`[vector-store] 已加载 ${vectors.length} 个向量`);
    }
  } catch (err) {
    console.warn('[vector-store] 加载向量文件失败:', err.message);
  }
  loaded = true;
}

/**
 * 持久化向量到磁盘
 */
function saveVectors() {
  if (!existsSync(VECTORS_DIR)) {
    mkdirSync(VECTORS_DIR, { recursive: true });
  }
  const filePath = join(VECTORS_DIR, 'vectors.json');
  writeFileSync(filePath, JSON.stringify(vectors), 'utf-8');
}

/**
 * 余弦相似度计算
 */
function cosineSimilarity(a, b) {
  let dot = 0, normA = 0, normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  const denom = Math.sqrt(normA) * Math.sqrt(normB);
  return denom === 0 ? 0 : dot / denom;
}

/**
 * 向量维度
 */
export const VECTOR_DIM = 384;

/**
 * 批量插入向量
 * @param {Array<{chunk_id: string, vector: number[]}>} items
 */
export function insertVectors(items) {
  loadVectors();

  // 去重
  const existing = new Set(vectors.map(v => v.chunk_id));
  for (const item of items) {
    if (existing.has(item.chunk_id)) {
      const idx = vectors.findIndex(v => v.chunk_id === item.chunk_id);
      vectors[idx] = item;
    } else {
      vectors.push(item);
      existing.add(item.chunk_id);
    }
  }

  saveVectors();
}

/**
 * 清空所有向量
 */
export function clearVectors() {
  vectors = [];
  saveVectors();
}

/**
 * 向量相似度检索
 * @param {number[]} queryVector - 查询向量
 * @param {number} topN - 返回数量
 * @returns {Array<{chunk_id: string, score: number}>}
 */
export function vectorSearch(queryVector, topN = 15) {
  loadVectors();

  if (vectors.length === 0) return [];

  // 计算所有相似度
  const scores = vectors.map(v => ({
    chunk_id: v.chunk_id,
    score: cosineSimilarity(queryVector, v.vector)
  }));

  // 排序取 top
  scores.sort((a, b) => b.score - a.score);

  return scores.slice(0, topN).map(r => ({
    chunk_id: r.chunk_id,
    score: Number(r.score.toFixed(4))
  }));
}

/**
 * 获取向量总数
 */
export function getVectorCount() {
  loadVectors();
  return vectors.length;
}

/**
 * 获取 chunk_id 对应的向量
 */
export function getVector(chunk_id) {
  loadVectors();
  const found = vectors.find(v => v.chunk_id === chunk_id);
  return found ? found.vector : null;
}

export default {
  VECTOR_DIM, insertVectors, clearVectors,
  vectorSearch, getVectorCount, getVector
};

File: rag/src/storage/vector_store.py
"""
vector_store.py — 向量存储层

基于 JSON 文件的向量存储，内存中管理，支持余弦相似度检索。
VECTOR_DIM = 384，与 sentence-transformers multilingual 模型对齐。
可替换为 Qdrant/Milvus 等专业向量数据库。
"""

import json
import math
import os
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent.parent
VECTORS_DIR = PROJECT_ROOT / "rag" / "data" / "vectors"
VECTORS_PATH = VECTORS_DIR / "vectors.json"

VECTOR_DIM = 384

# 内存存储: {chunk_id: {"vector": [...], "text": str, ...}}
_vectors = {}
_loaded = False


def _ensure_loaded():
    """确保向量数据已加载到内存"""
    global _loaded, _vectors
    if _loaded:
        return
    if VECTORS_PATH.exists():
        try:
            with open(VECTORS_PATH, "r", encoding="utf-8") as f:
                data = json.load(f)
                _vectors = data.get("vectors", {})
        except (json.JSONDecodeError, OSError) as e:
            print(f"[vector_store] 读取向量文件失败: {e}")
            _vectors = {}
    _loaded = True


def _save():
    """持久化到 JSON 文件"""
    VECTORS_DIR.mkdir(parents=True, exist_ok=True)
    with open(VECTORS_PATH, "w", encoding="utf-8") as f:
        json.dump({"vectors": _vectors, "dim": VECTOR_DIM}, f, ensure_ascii=False)


def cosine_similarity(vec_a, vec_b):
    """计算余弦相似度"""
    dot = 0.0
    norm_a = 0.0
    norm_b = 0.0
    for i in range(len(vec_a)):
        va = vec_a[i]
        vb = vec_b[i] if i < len(vec_b) else 0
        dot += va * vb
        norm_a += va * va
        norm_b += vb * vb
    denom = math.sqrt(norm_a) * math.sqrt(norm_b)
    if denom < 1e-10:
        return 0.0
    return dot / denom


def insert_vectors(chunk_vectors):
    """批量插入向量

    Args:
        chunk_vectors: list of {"chunk_id": str, "vector": list[float], "text": str}
    """
    _ensure_loaded()
    for item in chunk_vectors:
        _vectors[item["chunk_id"]] = {
            "vector": item["vector"],
            "text": item.get("text", ""),
        }
    _save()


def vector_search(query_vector, top_n=15):
    """余弦相似度搜索

    Args:
        query_vector: list[float]
        top_n: 返回数量

    Returns:
        list of {"chunk_id": str, "score": float}
    """
    _ensure_loaded()
    if not _vectors:
        return []

    scored = []
    for chunk_id, data in _vectors.items():
        score = cosine_similarity(query_vector, data["vector"])
        scored.append({"chunk_id": chunk_id, "score": score})

    scored.sort(key=lambda x: x["score"], reverse=True)
    return scored[:top_n]


def get_vector_count():
    """获取向量数量"""
    _ensure_loaded()
    return len(_vectors)


def clear():
    """清空所有向量"""
    global _vectors, _loaded
    _vectors = {}
    _loaded = True
    if VECTORS_PATH.exists():
        try:
            VECTORS_PATH.unlink()
        except OSError:
            pass
